src/systick.c

```diff
@@ -0,0 +1,82 @@
+#include "systick.h"
+#include "global_vars.h"
+#include "i2c.h"
+#include "adc.h"
+#include "output.h"
+#include "fan.h"
+#include "button.h"
+#include "led.h"
+#include "buzzer.h"
+
+#include <libopencm3/cm3/systick.h>
+#include <libopencm3/cm3/nvic.h>
+
+INA219_meas_t battery = {0};
+INA219_meas_t reg_5v = {0};
+int16_t board_temp = 0;
+
+uint8_t systick_slow_tick = 0;
+uint16_t systick_temp_tick = 0;
+uint8_t current_phase = 0;
+
+void systick_init(void) {
+    // Generate a 1ms systick interrupt
+    // 72MHz / 8 => 9000000 counts per second
+    systick_set_clocksource(STK_CSR_CLKSOURCE_AHB_DIV8);
+
+    // 9000000/9000 = 1000 overflows per second
+    // SysTick interrupt every N clock pulses
+    systick_set_reload(9000);
+
+    systick_interrupt_enable();
+
+    // Start counting.
+    systick_counter_enable();
+}
+
+void sys_tick_handler(void) {
+    // Every 20 ms read values from INA219 current sensors
+    if (++systick_slow_tick == 20) {
+        // if watchdog tripped re-init INA219's
+        if (i2c_timed_out) {
+            // reset watchdog
+            reset_i2c_watchdog();
+            init_i2c_sensors();
+        }
+        // Read INA219's
+        battery = measure_current_sense(BATTERY_SENSE_ADDR);
+        battery.current *= 10;  // convert to 1mA LSB
```

Comment:
Is there a risk of overflow here? The current field is int16, so can represent a maximum of ~32A. I can't remember the maximum current that can be measured with the INA219, but I suspect it might be more than this?

Also, any risk of concurrency issues here?

Reply:
Good catch! I'd adjusted the INA219 to be 10 mA LSB to avoid the overflow on the INA219 end but forgot to increase the int size I was storing it in. It now won't overflow until about 320A and the INA219 can measure up to ±160A so we should be good there.

What do you mean in terms of concurrency? After startup the I2C is only accessed from the systick ISR so is all done serially and blocking. Same is true of the ADC and current multiplexing lines.

Reply:
Regarding concurrency, I was referring to the non-atomic setting of the battery variable. If this is the highest priority/only IRQ then it's fine as nothing can interrupt between the initial assignment to `battery` and its modification. 

Another thing that's just occurred to me is that it may be necessary to set the battery variable as volatile (same goes for other variables modified in an ISR), otherwise the optimiser could mess things up.

Reply:
Yeah, other than the watchdog the systick is the only ISR in use. I got a bit sloppy with setting volatile, have corrected.